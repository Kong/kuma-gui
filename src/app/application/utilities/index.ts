type URLParamDefault = string | number | boolean
type URLParamValue = string | null

const difference = <T>(a: T[], b: T[]): T[] => a.filter(item => !b.includes(item))

const includes = <T extends readonly string[]>(arr: T, item: string): item is T[number] => {
  return arr.includes(item as T[number])
}

const createUniqueId = (j = 0) => {
  let i = j
  return (prefix = 'unique') => {
    i++
    return `${prefix}-${i}`
  }
}

export const uniqueId = createUniqueId()

export const beforePaint = <T extends (...args: any[]) => void>(fn: T) => {
  let num: number
  return (...args: Parameters<T>) => {
    if (num) {
      window.cancelAnimationFrame(num)
    }
    num = window.requestAnimationFrame(fn.bind(fn, ...args))
  }
}

const supportedAttrs = ['class'] as const
type SupportedAttrs = typeof supportedAttrs[number]
export const createAttrsSetter = ($el = document.documentElement) => {
  if (!$el) {
    return () => {}
  }
  const originalClasses = [...$el.classList]
  return beforePaint((attrs: Partial<Record<SupportedAttrs, string>>[]) => {
    const flat = attrs.reduce<Record<SupportedAttrs, string[]>>((prev, item) => {
      return Object.entries(item).reduce(
        (prev, [key, value]) => {
          if (includes(supportedAttrs, key) && value) {
            prev[key].push(value)
          }
          return prev
        }, prev,
      )
    }, { class: [] })

    // omit any classes that were on the node previous to our application starting
    const currentClasses = difference([...$el.classList], originalClasses)
    // anything in currentClasses that isn't in our tree of attrs, remove
    $el.classList.remove(...difference(currentClasses, flat.class))
    // anything in our tree of attrs that isn't in currentClasses, add
    $el.classList.add(...difference(flat.class, currentClasses))
  })
}

// normalizes url params from "value or array of values" to value
// if the url params is an array we use the first value
export const urlParam = function <T extends URLParamValue> (param: T | T[]): T {
  return (Array.isArray(param) ? param[0] : param)
}

//
export const normalizeUrlParam = (param: URLParamValue, def: URLParamDefault): URLParamDefault => {
  switch (true) {
    case typeof def === 'boolean':
      return param === null ? true : def
    case typeof def === 'number': {
      const value = param === null || param.length === 0 ? def : Number(decodeURIComponent(param))
      if (isNaN(value)) {
        return Number(def)
      } else {
        return value
      }
    }
    case typeof def === 'string': {
      return param === null || param.length === 0 ? def : decodeURIComponent(param)
    }
  }
  throw new TypeError('URL parameters can only be string | number | boolean')
}
export const createTitleSetter = ($doc = document) => {
  return beforePaint((title) => {
    $doc.title = title
  })
}
// when used with router.push, prevents things like `q=` (i.e. key= but with no value)
export const cleanQuery = <T extends Record<string, unknown>>(params: Record<string, string | boolean | number | undefined>, originalQuery: T) => {
  const query = {
    ...originalQuery as Record<string, string | undefined | null>,
  }
  const processed = Object.entries(params).reduce((prev, [key, value]) => {
    // Only add parameters for non-empty strings or non-false boolean
    switch (true) {
      case value === true:
        // if the value is strictly true that means we use shorthand boolean true
        // i.e. `?prop`
        prev[key] = null
        break
      case value !== false && String(value).length > 0:
        // if we aren't strictly false (which will use the default)
        // convert to a string and check the length
        // anything greater than zero encode and use
        prev[key] = encodeURIComponent(String(value))
        break
      default:
        // false, or any zero length value unset entirely
        prev[key] = undefined
    }
    return prev
  }, query)
  return {
    ...query,
    ...processed,
  }
}
