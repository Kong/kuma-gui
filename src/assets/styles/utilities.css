/* Utilities */

.kutil-sr-only {
  position: absolute !important;
  width: 1px !important;
  height: 1px !important;
  padding: 0 !important;
  margin: -1px !important;
  overflow: hidden !important;
  clip: rect(0, 0, 0, 0) !important;
  white-space: nowrap !important;
  border: 0 !important;
}

.kutil-img-full {
  max-width: 100%;
  width: 100%;
  height: auto;
}

.kutil-img-fluid {
  max-width: 100%;
  height: auto;
}

.kutil-clear-text {
  -webkit-font-smoothing: antialiased;
  text-rendering: optimizeLegibility;
  -moz-osx-font-smoothing: grayscale;
}

.kutil-button-base {
  font-family: $base-font-family;
  font-weight: $fw-medium;
  color: #fff !important;
  background-color: $green-base;
  padding: 0.563rem 2.25rem;
  border-radius: 3px;
  box-shadow: $base-soft-shadow;
  text-align: center;
  text-shadow: 0 2px 0 $green-dark;
  box-sizing: border-box;
  cursor: pointer;
}

.kutil-button-base:hover,
.kutil-button-base:active,
.kutil-button-base:focus,
.kutil-button-base:visited {
  color: #fff !important;
}

.kutil-button-base:hover {
  /* background-color: lighten($green-base, 3.5); */
}

.kutil-quotify {
  quotes: "“" "”" "‘" "’";
}

.kutil-quotify:before {
  content: open-quote;
}

.kutil-quotify:after {
  content: close-quote;
}

/* @mixin retina-mq {
  @media (-webkit-min-device-pixel-ratio: 2), (min-resolution: 192dpi) {
    @content;
  }
}

@mixin target-ie {
  @media all and (-ms-high-contrast: none), (-ms-high-contrast: active) {
    @content;
  }
} */

.kutil-sticky {
  position: -webkit-sticky;
  position: sticky;
  top: 0;
}

.kutil-sticky:before,
.kutil-sticky:after {
  content: "";
  display: table;
}

/* New for Kuma */

.arrow-link {
  display: flex;
  align-items: center;
  text-decoration: none;
  color: var(--blue-4);
}

.k-button {
  font-weight: normal !important;
  display: flex;
  align-items: center;
}

.k-button:after,
.arrow-link:after {
  display: inline-block;
  content: "›";
  margin-left: 5px;
}

.capitalize {
  display: inline-block;
  text-transform: capitalize;
}
