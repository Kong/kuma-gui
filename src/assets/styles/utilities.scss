/* Utilities */

.kutil-sr-only {
  position: absolute !important;
  width: 1px !important;
  height: 1px !important;
  padding: 0 !important;
  margin: -1px !important;
  overflow: hidden !important;
  clip: rect(0, 0, 0, 0) !important;
  white-space: nowrap !important;
  border: 0 !important;
}

.kutil-img-full {
  max-width: 100%;
  width: 100%;
  height: auto;
}

.kutil-img-fluid {
  max-width: 100%;
  height: auto;
}

.kutil-clear-text {
  -webkit-font-smoothing: antialiased;
  text-rendering: optimizeLegibility;
  -moz-osx-font-smoothing: grayscale;
}

.kutil-quotify {
  quotes: "“" "”" "‘" "’";
}

.kutil-quotify:before {
  content: open-quote;
}

.kutil-quotify:after {
  content: close-quote;
}

/* @mixin retina-mq {
  @media (-webkit-min-device-pixel-ratio: 2), (min-resolution: 192dpi) {
    @content;
  }
}

@mixin target-ie {
  @media all and (-ms-high-contrast: none), (-ms-high-contrast: active) {
    @content;
  }
} */

.kutil-sticky {
  position: -webkit-sticky;
  position: sticky;
  top: 0;
}

.kutil-sticky:before,
.kutil-sticky:after {
  content: "";
  display: table;
}

/* New for Kuma */

.capitalize {
  display: inline-block;
  text-transform: capitalize;
}

/* position */
.fixed-top {
  position: fixed;
  top: 0;
  right: 0;
  left: 0;
  z-index: 1030;
}

.text-left {
  text-align: left;
}

.text-center {
  text-align: center;
}

.text-right {
  text-align: right;
}

.text-uppercase {
  text-transform: uppercase;
}

.text-italic {
  font-style: italic;
}

.font-normal {
  font-weight: 400;
}

.font-light {
  font-weight: 100;
}

.font-semibold {
  font-weight: 500;
}

.font-bold {
  font-weight: 700;
}

/* badges */
.badge {
  display: inline-block;
  padding: 0.25rem 0.4rem;
  font-size: 75%;
  line-height: 1;
  text-align: center;
  white-space: nowrap;
  vertical-align: baseline;
  border-radius: 0.25rem;
}

.top-border {
  padding-top: 1.5rem;
  border-top: 1px solid var(--black-10);
}

.bottom-border {
  padding-bottom: 1.5rem;
  border-bottom: 1px solid var(--black-10);
}

.text-muted {
  color: var(--black-45) !important;
}
.text-danger {
  color: var(--red-500) !important;
}

.h-100 {
  height: 100%;
}
.h-screen {
  min-height: 100vh;
}

code.inline {
  @apply py-1 px-2 m-0 text-sm rounded;

  background-color: var(--gray-1);
}

code.block {
  @apply my-4 rounded-md overflow-auto p-4 text-sm;

  background-color: var(--gray-1);
  border: 1px solid var(--gray-4);
}
