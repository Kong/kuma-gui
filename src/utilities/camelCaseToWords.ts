export function camelCaseToWords(str: string): string {
  const words = str
    .split(/([A-Z][a-z]+)/)
    .join(' ')
    .replace(/\s+/g, ' ')
    .trim()
  return words.charAt(0).toUpperCase() + words.substring(1)
}
