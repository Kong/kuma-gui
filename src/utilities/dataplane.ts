import {
  Compatibility,
  DataPlaneInsight,
  DataPlaneNetworking,
  DataPlaneOverview,
  DiscoverySubscription,
  KDSSubscription,
  LabelValue,
  StatusKeyword,
  Version,
} from '@/types/index.d'

/**
 * Takes a data plane and constructs the list of tags. It removes duplicate tags so we don't display them twice. Note that tags are only considered a duplicate if both their key and their value are the same.
 *
 * **Example**:
 *
 * Data plane:
 *
 * ```yaml
 * type: Dataplane
 * mesh: default
 * name: cluster-1.ingress-01
 * networking:
 *   inbound:
 *     - port: 1234
 *       tags:
 *         kuma.io/service: backend
 *         version: 1
 *     - port: 1235
 *       tags:
 *         kuma.io/service: backend-api
 *         version: 1
 * ```
 *
 * Output:
 *
 * ```js
 * [
 *   { label: 'kuma.io/service', value: 'backend'},
 *   { label: 'kuma.io/service', value: 'backend-api'},
 *   { label: 'version', value: '1'},
 * ]
 * ```
 */
export function dpTags(dataplane: { networking: DataPlaneNetworking }): LabelValue[] {
  let tags: string[] = []

  if (dataplane.networking.inbound) {
    tags = dataplane.networking.inbound
      .filter((inbound) => 'tags' in inbound)
      .flatMap((inbound) => Object.entries(inbound.tags))
      .map(([key, value]) => `${key}=${value}`)
  }

  if (dataplane.networking.gateway) {
    // gateway data plane has no inbounds, but has tags embedded in gateway branch
    tags = Object.entries(dataplane.networking.gateway.tags).map(([key, value]) => `${key}=${value}`)
  }

  const uniqueTags = Array.from(new Set(tags))

  uniqueTags.sort((tagPairA, tagPairB) => tagPairA.localeCompare(tagPairB))

  return uniqueTags
    .map((tagPair) => tagPair.split('='))
    .map(([label, value]) => ({ label, value }))
}

// getItemStatusFromInsight takes object with subscriptions and returns a
// status 'online' | 'offline'
export function getItemStatusFromInsight(insight: { subscriptions?: DiscoverySubscription[] | KDSSubscription[] } | undefined = { subscriptions: [] }): StatusKeyword {
  const proxyOnline = (insight.subscriptions ?? []).some((subscription) => subscription.connectTime?.length && !subscription.disconnectTime)
  return proxyOnline ? 'online' : 'offline'
}

// getStatusAndReason takes Dataplane and DataplaneInsight and returns a
// {status: 'online' | 'offline' | 'partially_degraded', reason: errors[]}
export function getStatusAndReason(dataplane: { networking: DataPlaneNetworking }, insight: { subscriptions?: DiscoverySubscription[] } | undefined = { subscriptions: [] }): { status: StatusKeyword, reason: string[] } {
  const inbound = dataplane.networking.inbound ?? []
  const errors = inbound
    .filter(item => item.health && !item.health.ready)
    .map(item => `Inbound on port ${item.port} is not ready (kuma.io/service: ${item.tags['kuma.io/service']})`)

  let status: StatusKeyword
  switch (true) {
    case inbound.length === 0:
      status = 'online'
      break
    // if errors and inbounds are equal, even if they are both 0
    // then we are offline
    case errors.length === inbound.length:
      status = 'offline'
      break
    // otherwise any errors at all, we are degraded
    case errors.length > 0:
      status = 'partially_degraded'
      break
    default:
      // otherwise run the normal getter
      status = getItemStatusFromInsight(insight)
  }
  return {
    status,
    reason: errors,
  }
}

/*
getStatus takes DataplaneInsight and returns map of versions
 */

export function getVersions(dataPlaneInsight: DataPlaneInsight | undefined): Record<string, string> | null {
  const subscriptions = dataPlaneInsight?.subscriptions || []
  if (subscriptions.length === 0) {
    return null
  }

  const versions: Record<string, string> = {}

  const lastSubscription = subscriptions[subscriptions.length - 1]

  if (lastSubscription.version === undefined) {
    return null
  }

  if (lastSubscription.version.envoy) {
    versions.envoy = lastSubscription.version.envoy.version
  }

  if (lastSubscription.version.kumaDp) {
    versions.kumaDp = lastSubscription.version.kumaDp.version
  }

  if (lastSubscription.version.dependencies) {
    Object.entries(lastSubscription.version.dependencies).forEach(([key, value]) => {
      versions[key] = value
    })
  }

  return versions
}

export function parseMTLSData(dataPlaneOverview: DataPlaneOverview, formatIsoDate: (isoDate: string) => string) {
  if (dataPlaneOverview.dataplaneInsight === undefined || dataPlaneOverview.dataplaneInsight.mTLS === undefined) {
    return null
  }

  const { mTLS } = dataPlaneOverview.dataplaneInsight
  const rawExpDate = new Date(mTLS.certificateExpirationTime)
  // this prevents any weird date shifting
  const fixedExpDate = new Date(rawExpDate.getTime() + rawExpDate.getTimezoneOffset() * 60000)

  return {
    certificateExpirationTime: formatIsoDate(fixedExpDate.toISOString()),
    lastCertificateRegeneration: formatIsoDate(mTLS.lastCertificateRegeneration),
    certificateRegenerations: mTLS.certificateRegenerations,
  }
}

export function compatibilityKind(version: Version): Compatibility {
  const isKumaCpCompatible = version.kumaDp?.kumaCpCompatible ?? true

  if (!isKumaCpCompatible) {
    return {
      kind: INCOMPATIBLE_UNSUPPORTED_KUMA_DP,
      payload: {
        kumaDp: version.kumaDp.version,
      },
    }
  }

  const isKumaDpCompatible = version.envoy?.kumaDpCompatible ?? true

  if (!isKumaDpCompatible) {
    return {
      kind: INCOMPATIBLE_UNSUPPORTED_ENVOY,
      payload: {
        envoy: version.envoy.version,
        kumaDp: version.kumaDp.version,
      },
    }
  }

  return {
    kind: COMPATIBLE,
  }
}

export const COMPATIBLE = 'COMPATIBLE'
export const INCOMPATIBLE_ZONE_CP_AND_KUMA_DP_VERSIONS = 'INCOMPATIBLE_ZONE_CP_AND_KUMA_DP_VERSIONS'
export const INCOMPATIBLE_ZONE_AND_GLOBAL_CPS_VERSIONS = 'INCOMPATIBLE_ZONE_AND_GLOBAL_CPS_VERSIONS'
export const INCOMPATIBLE_UNSUPPORTED_KUMA_DP = 'INCOMPATIBLE_UNSUPPORTED_KUMA_DP'
export const INCOMPATIBLE_UNSUPPORTED_ENVOY = 'INCOMPATIBLE_UNSUPPORTED_ENVOY'
export const INCOMPATIBLE_WRONG_FORMAT = 'INCOMPATIBLE_WRONG_FORMAT'
