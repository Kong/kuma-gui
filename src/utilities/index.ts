import { TOKENS, createInjections } from '@/services'

export const [
  useEnv,
  useNav,
  useKumaApi,
  useStore,
  useRouter,
] = createInjections(TOKENS.env, TOKENS.nav, TOKENS.api, TOKENS.store, TOKENS.router)
