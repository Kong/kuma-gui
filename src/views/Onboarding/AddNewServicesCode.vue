<template>
  <OnboardingPage>
    <template #header>
      <OnboardingHeading title="Add services" />
    </template>
    <template #content>
      <p class="text-center mb-12">
        The demo application includes two services: a Redis backend to store a counter value,
        and a frontend web UI to show and increment the counter.
      </p>

      <template v-if="isKubernetes">
        <p>To run execute the following command:</p>

        <CodeBlock
          language="bash"
          :code="k8sRunCommand"
        />
      </template>

      <div v-else>
        <p>Clone the GitHub repository for the demo application:</p>

        <CodeBlock
          language="bash"
          :code="githubLink"
        />

        <KCard
          title="And follow the instructions in the README"
          border-variant="noBorder"
        >
          <template #body>
            <a
              target="_blank"
              class="external-link-code-block"
              :href="githubLinkReadme"
            >
              {{ githubLinkReadme }}
            </a>
          </template>
        </KCard>
      </div>
      <div>
        <p class="text-center my-4">
          DPPs status:
          <span
            v-if="hasDPPs"
            class="text-green-500"
            data-testid="dpps-connected"
          >Connected</span>
          <span
            v-else
            class="text-red-500"
            data-testid="dpps-disconnected"
          >Disconeccted</span>
        </p>
        <div
          v-if="!hasDPPs"
          class="flex justify-center"
        >
          <LoadingBox />
        </div>
      </div>
    </template>
    <template #navigation>
      <OnboardingNavigation
        next-step="onboarding-dataplanes-overview"
        previous-step="onboarding-add-services"
        :should-allow-next="hasDPPs"
      />
    </template>
  </OnboardingPage>
</template>

<script>
import { mapGetters } from 'vuex'
import json2yaml from '@appscode/json2yaml'
import { KCard } from '@kong/kongponents'

import Kuma from '@/services/kuma'
import { PRODUCT_NAME } from '@/consts'
import { kumaDpServerUrl } from '@/configUrl'
import CodeBlock from '@/components/CodeBlock.vue'
import LoadingBox from '@/components/LoadingBox.vue'
import OnboardingNavigation from '@/views/Onboarding/components/OnboardingNavigation.vue'
import OnboardingHeading from '@/views/Onboarding/components/OnboardingHeading.vue'
import OnboardingPage from '@/views/Onboarding/components/OnboardingPage.vue'

const LONG_POOLING_INTERVAL = 1000

export default {
  name: 'AddNewServicesCode',
  components: {
    CodeBlock,
    OnboardingNavigation,
    OnboardingHeading,
    OnboardingPage,
    LoadingBox,
    KCard,
  },
  metaInfo() {
    return {
      title: 'Add new services',
    }
  },
  data() {
    return {
      productName: PRODUCT_NAME,
      githubLink: 'https://github.com/kumahq/kuma-counter-demo/',
      githubLinkReadme: 'https://github.com/kumahq/kuma-counter-demo/blob/master/README.md',
      k8sRunCommand: 'kubectl apply -f https://bit.ly/3Kh2Try',
      generateDpTokenCode: 'kumactl generate dataplane-token --name=redis > kuma-token-redis',
      startDpCode: `kuma-dp run \\
      --cp-address=${kumaDpServerUrl()} \\
      --dataplane=${`"${json2yaml({
    type: 'Dataplane',
    mesh: 'default',
    name: 'example',
    networking: {
      address: 'localhost',
      inbound: [
        {
          port: 7777,
          servicePort: 7777,
          serviceAddress: '127.0.0.1',
          tags: {
            'kuma.io/service': 'example',
            'kuma.io/protocol': 'tcp',
          },
        },
      ],
    },
  })}"`} \\
      --dataplane-token-file=kuma-token-example`,
      hasDPPs: false,
      DPPsTimeout: null,
    }
  },
  computed: {
    ...mapGetters({
      environment: 'config/getEnvironment',
    }),
    isKubernetes() {
      return this.environment === 'kubernetes'
    },
  },
  created() {
    this.getDPPs()
  },
  unmounted() {
    clearTimeout(this.DPPsTimeout)
  },

  methods: {
    async getDPPs() {
      try {
        const { total } = await Kuma.getAllDataplanes()

        this.hasDPPs = total > 0
      } catch (e) {
        console.error(e)
      }

      if (!this.hasDPPs) {
        this.DPPsTimeout = setTimeout(() => this.getDPPs(), LONG_POOLING_INTERVAL)
      }
    },
  },
}
</script>
