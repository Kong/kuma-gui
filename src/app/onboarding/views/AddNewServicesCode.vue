<template>
  <RouteView>
    <RouteTitle
      :title="t('onboarding.routes.add-services-code.title')"
    />
    <AppView>
      <OnboardingPage>
        <template #header>
          <OnboardingHeading>
            <template #title>
              Add services
            </template>
          </OnboardingHeading>
        </template>

        <template #content>
          <p class="mb-4 text-center">
            The demo application includes two services: a Redis backend to store a counter value, and a frontend web UI to show and increment the counter.
          </p>

          <template v-if="isKubernetes">
            <p>To run execute the following command:</p>

            <CodeBlock
              id="code-block-kubernetes-command"
              language="bash"
              :code="k8sRunCommand"
            />
          </template>

          <div v-else>
            <p class="mb-4 text-center">
              Clone <a
                :href="githubLink"
                target="_blank"
              >the GitHub repository</a> for the demo application:
            </p>

            <CodeBlock
              id="code-block-clone-command"
              language="bash"
              :code="`git clone ${githubLink}`"
            />

            <p class="mt-4 text-center">
              And follow the instructions in <a
                :href="githubLinkReadme"
                target="_blank"
              >the README</a>.
            </p>
          </div>

          <div>
            <p class="status-box mt-4">
              DPPs status:

              <span
                v-if="hasDPPs"
                class="status--is-connected"
                data-testid="dpps-connected"
              >Connected</span>

              <span
                v-else
                class="status--is-disconnected"
                data-testid="dpps-disconnected"
              >Disconnected</span>
            </p>

            <div
              v-if="!hasDPPs"
              class="status-loading-box mt-4"
            >
              <LoadingBox />
            </div>
          </div>
        </template>

        <template #navigation>
          <OnboardingNavigation
            next-step="onboarding-dataplanes-overview"
            previous-step="onboarding-add-services"
            :should-allow-next="hasDPPs"
          />
        </template>
      </OnboardingPage>
    </AppView>
  </RouteView>
</template>

<script lang="ts" setup>
import { computed, onUnmounted, ref } from 'vue'

import LoadingBox from '../components/LoadingBox.vue'
import OnboardingHeading from '../components/OnboardingHeading.vue'
import OnboardingNavigation from '../components/OnboardingNavigation.vue'
import OnboardingPage from '../components/OnboardingPage.vue'
import AppView from '@/app/application/components/app-view/AppView.vue'
import RouteTitle from '@/app/application/components/route-view/RouteTitle.vue'
import RouteView from '@/app/application/components/route-view/RouteView.vue'
import CodeBlock from '@/app/common/CodeBlock.vue'
import { useStore } from '@/store/store'
import { useKumaApi, useI18n } from '@/utilities'
const { t } = useI18n()

const kumaApi = useKumaApi()
const store = useStore()

const LONG_POOLING_INTERVAL = 1000

const githubLink = 'https://github.com/kumahq/kuma-counter-demo/'
const githubLinkReadme = 'https://github.com/kumahq/kuma-counter-demo/blob/master/README.md'
const k8sRunCommand = 'kubectl apply -f https://bit.ly/3Kh2Try'

const hasDPPs = ref(false)
const dppsTimeout = ref<number | null>(null)

const isKubernetes = computed(() => store.getters['config/getEnvironment'] === 'kubernetes')

getDPPs()

onUnmounted(function () {
  clearTimeout()
})

async function getDPPs() {
  try {
    const { total } = await kumaApi.getAllDataplanes()

    hasDPPs.value = total > 0
  } catch (err) {
    console.error(err)
  } finally {
    if (!hasDPPs.value) {
      clearTimeout()
      dppsTimeout.value = window.setTimeout(() => getDPPs(), LONG_POOLING_INTERVAL)
    }
  }
}

function clearTimeout(): void {
  if (dppsTimeout.value !== null) {
    window.clearTimeout(dppsTimeout.value)
  }
}
</script>

<style lang="scss" scoped>
.status-box {
  text-align: center;
}

.status--is-connected {
  color: $kui-color-text-success;
}

.status--is-disconnected {
  color: $kui-color-text-danger;
}

.status-loading-box {
  display: flex;
  justify-content: center;
}
</style>
