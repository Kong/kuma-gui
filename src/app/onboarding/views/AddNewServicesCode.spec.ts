import { describe, expect, jest, test } from '@jest/globals'
import { flushPromises, mount } from '@vue/test-utils'

import AddNewServicesCode from './AddNewServicesCode.vue'
import { useKumaApi } from '@/utilities'

const kumaApi = useKumaApi()

function renderComponent() {
  return mount(AddNewServicesCode)
}

describe('AddNewServicesCode.vue', () => {
  test('renders snapshot', async () => {
    const wrapper = renderComponent()

    await flushPromises()

    expect(wrapper.element).toMatchSnapshot()
  })

  test('detects resources on call and allow to proceed', async () => {
    const wrapper = renderComponent()

    await flushPromises()

    expect(wrapper.find('[data-testid="dpps-connected"]').exists()).toBe(true)
    expect(wrapper.find('[data-testid="onboarding-next-button"]').exists()).toBe(true)
    expect(wrapper.find('[data-testid="loading"]').exists()).toBe(false)
  })

  test('refetch resources if any not available', async () => {
    jest.useFakeTimers()
    jest
      .spyOn(kumaApi, 'getAllDataplanes')
      .mockResolvedValueOnce({
        items: [],
        total: 0,
        next: null,
      })
      .mockResolvedValueOnce({
        items: [],
        total: 1,
        next: null,
      })

    const wrapper = renderComponent()

    expect(wrapper.find('[data-testid="loading"]').exists()).toBe(true)
    expect(wrapper.find('[data-testid="dpps-disconnected"]').exists()).toBe(true)

    await flushPromises()
    jest.runAllTimers()
    await flushPromises()

    expect(wrapper.find('[data-testid="dpps-connected"]').exists()).toBe(true)
    expect(kumaApi.getAllDataplanes).toHaveBeenCalledTimes(2)

    jest.useRealTimers()
  })
})
