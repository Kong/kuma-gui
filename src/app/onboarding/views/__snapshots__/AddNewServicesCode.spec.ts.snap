// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddNewServicesCode.vue renders snapshot 1`] = `
<div
  class="route-view"
>
  <div
    aria-atomic="true"
    aria-live="assertive"
    class="route-view-title visually-hidden"
  >
    Navigated to Add new services
  </div>
  
  <!--v-if-->
  <main
    class="app-main-content"
  >
    
    <!--v-if-->
    <section
      class=""
    >
      <!--v-if-->
      
      <div>
        <div
          class="onboarding-container"
        >
          <div
            class="onboarding-container__header"
          >
            
            <div
              class="onboarding-heading"
            >
              <h1
                class="onboarding-title"
              >
                
                 Add services 
                
              </h1>
              <!--v-if-->
            </div>
            
          </div>
          <div
            class="onboarding-container__content"
          >
            <div
              class="onboarding-container__inner-content"
            >
              
              <p
                class="mb-4 text-center"
              >
                 The demo application includes two services: a Redis backend to store a counter value, and a frontend web UI to show and increment the counter. 
              </p>
              <div>
                <p
                  class="mb-4 text-center"
                >
                   Clone 
                  <a
                    href="https://github.com/kumahq/kuma-counter-demo/"
                    target="_blank"
                  >
                    the GitHub repository
                  </a>
                   for the demo application: 
                </p>
                <div
                  class="k-code-block theme-dark code-block"
                  data-testid="k-code-block"
                  id="code-block-clone-command"
                  style="--maxLineNumberWidth: 1ch;"
                  tabindex="0"
                >
                  <!---->
                  <div
                    class="k-code-block-content"
                  >
                    <pre
                      class="k-highlighted-code-block show-copy-button language-bash"
                      data-testid="k-code-block-highlighted-code-block"
                      tabindex="0"
                    >
                              
                      <span
                        class="k-line-number-rows"
                      >
                        
          
                        
                        <span
                          class="k-line"
                        >
                          
            
                          <a
                            class="k-line-anchor"
                            id="code-block-clone-command-L1"
                          >
                            1
                          </a>
                          
          
                        </span>
                        
                        
        
                      </span>
                      
        
                      <code
                        class="language-bash"
                      >
                        <span
                          class="token function"
                        >
                          git
                        </span>
                         clone https://github.com/kumahq/kuma-counter-demo/
                      </code>
                      
      
                    </pre>
                    <div
                      class="k-code-block-secondary-actions"
                    >
                      <button
                        class="k-button small outline k-code-block-copy-button"
                        data-testid="k-code-block-copy-button"
                        title="Copy (Alt+C)"
                        type="button"
                      >
                        
                        <!---->
                        
                        
                        <span
                          class="kong-icon kong-icon-copy"
                        >
                          <svg
                            height="16px"
                            role="img"
                            viewBox="0 0 32 32"
                            width="16px"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            
  
                            <title>
                              Copy (Alt+C)
                            </title>
                            
  
                            <path
                              d="M25.313 28v-18.688h-14.625v18.688h14.625zM25.313 6.688c1.438 0 2.688 1.188 2.688 2.625v18.688c0 1.438-1.25 2.688-2.688 2.688h-14.625c-1.438 0-2.688-1.25-2.688-2.688v-18.688c0-1.438 1.25-2.625 2.688-2.625h14.625zM21.313 1.313v2.688h-16v18.688h-2.625v-18.688c0-1.438 1.188-2.688 2.625-2.688h16z"
                              fill="currentColor"
                            />
                            

                          </svg>
                          

                        </span>
                        <span
                          class="visually-hidden"
                        >
                          Copy
                        </span>
                        
                        <!---->
                      </button>
                      
                      
                    </div>
                  </div>
                </div>
                <p
                  class="mt-4 text-center"
                >
                   And follow the instructions in 
                  <a
                    href="https://github.com/kumahq/kuma-counter-demo/blob/master/README.md"
                    target="_blank"
                  >
                    the README
                  </a>
                  . 
                </p>
              </div>
              <div>
                <p
                  class="status-box mt-4"
                >
                   DPPs status: 
                  <span
                    class="status--is-connected"
                    data-testid="dpps-connected"
                  >
                    Connected
                  </span>
                </p>
                <!--v-if-->
              </div>
              
            </div>
          </div>
          <div
            class="mt-4"
          >
            
            <div
              class="onboarding-actions"
            >
              <a
                class="k-button medium rounded secondary"
                data-testid="onboarding-previous-button"
                href="/gui/onboarding/add-services"
                type="button"
              >
                
                <!---->
                
                
                 Back 
                
                <!---->
              </a>
              <div
                class="button-list"
              >
                <a
                  aria-current="page"
                  class="router-link-active router-link-exact-active k-button medium rounded outline"
                  data-testid="onboarding-skip-button"
                  href="/gui/"
                  type="button"
                >
                  
                  <!---->
                  
                  
                   Skip setup 
                  
                  <!---->
                </a>
                <a
                  class="k-button medium rounded primary"
                  data-testid="onboarding-next-button"
                  href="/gui/onboarding/dataplanes-overview"
                  type="button"
                >
                  
                  <!---->
                  
                  
                  Next
                  
                  <!---->
                </a>
              </div>
            </div>
            
          </div>
        </div>
        <div
          class="background-image"
        />
      </div>
      
    </section>
    
  </main>
  
</div>
`;
